### PR TITLE
v.0.0.3

### PR DESCRIPTION
**Pull request by kayliedehart**
- Support for ApiFilters
- Support for Pagination (with ApiFilters)
- Generic delete  by id or slug
- fetch all Notes for a target slug

(in progress: unit testing for `mock-cloud`)

### CODE REVIEW

**Review by timburke — APPROVED** [review]
Do we not have any tests on `ts-iotile-cloud` that would break now that we've changed the way the fetch request GETs are structured?  We used to have tests with a mock http backend.

**Comment by kayliedehart** [comment]
@timburke I'll add some - currently we're setting up the mock cloud in `ts-iotile-cloud`, but only calling it in the [external] tests where it's used (Cache Service, etc) which isn't really in the spirit of unit tests.

File: src/cloud/iotile-cloud-serv.ts [hunk 1]
```diff
@@ -357,11 +357,10 @@ export class IOTileCloud {
    *
    * @returns {Device[]} A list of the IOTile devices.
    */
-  // TODO: get by pages of 500
   public async fetchAllDevices() {
     let that = this;
     return new Promise<Device[]>(function(resolve, reject) {
-      that.fetchFromServer('/device/?page_size=1000') //The default limit of devices returned per page is apparently 100
+      that.fetchPagesFromServer('/device/', 500)
```
Comment (dkarchmer) [inline src/cloud/iotile-cloud-serv.ts]: You may want to look into https://github.com/iotile/ng-iotile-cloud/blob/master/src/models/apifilter.ts and enhance the fetchPageFromService not to take 500, but to take an apiFilter, allowing you to reuse this function to filter a lot more. You can see how the apiFilter is use as an optional argument e.g. https://github.com/iotile/ng-iotile-cloud/blob/master/src/cloud.service.ts#L254

File: src/cloud/iotile-cloud-serv.ts [hunk 2]
```diff
@@ -588,6 +586,30 @@ export class IOTileCloud {
     });
   }
 
+  public async deleteStreamData(dataId: number) {
```
Comment (dkarchmer) [inline src/cloud/iotile-cloud-serv.ts]: I wonder if we should instead have a more generic `delete(resource: string, id: number)` where resource is either `note` or `data` or anything else that supports delete

File: src/cloud/iotile-cloud-serv.ts [hunk 3]
```diff
@@ -629,6 +651,24 @@ export class IOTileCloud {
     });
   }
 
+
+  public async fetchNotes(targetSlug: string): Promise<Array<Note>> {
```
Comment (dkarchmer) [inline src/cloud/iotile-cloud-serv.ts]: As described above, please add the apiFilter so we can use this function with other filters (i.e. date ranges)

File: src/cloud/iotile-cloud-serv.ts [hunk 4]
```diff
@@ -1110,6 +1151,84 @@ export class IOTileCloud {
     });
   }
 
+  private async fetchPagesFromServer(uri: string, pageSize: number, headers?: {}) : Promise<{} | Array<{}>> {
```
Comment (dkarchmer) [inline src/cloud/iotile-cloud-serv.ts]: Add apiFilter

**Review by dkarchmer — APPROVED** [review]
I think this would be a good time to make the API a little more generic, so we do not need to make changes to the API every time we want to filter or change the page_size or not have a page_size. Instead, I propose we do something similar to what we did for ng-iotile-cloud where we used a simple model apiFilter that basically just knows how to format parameters into the argument string. Should be an easy enhancement and will make the API more powerful.

I am approving in case you want to put this back now and do the change later, but I strongly suggest this enhancement is made. A page_size limit should really NOT be defined (even as a default) on reusable library

**Comment by kayliedehart** [comment]
@timburke @dkarchmer Made changes to support optional ApiFilter arguments where I thought it might make sense. Generic delete, and refactored fetchPagesFromServer to just be called by the original fetchFromServer to deduplicate code. 

I've tested the packages changes locally w/ Companion && tally.iotile.cloud (w/ the appropriate tweaks made), but I'm still working on real tests in `mock-cloud.spec.ts`

**Comment by dkarchmer** [comment]
Looks good. thanks

**Pull request merged**